main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,243 @@
+    }
+
+    def incrementLocalDefCount: OccurInfo = {
+      this.copy(localDefs = localDefs + 1)
+    }
+  }
+
+  private def increment(occurInfo: OccurInfo): OccurInfo = {
+    occurInfo.copy(size = occurInfo.size + 1)
+  }
+
+  /**
+    * A set of functions that contain masked casts.
+    *
+    * Must be manually maintained since Lowering erases the masked cast.
+    */
+  private val DangerousFunctions: Set[String] = Set("bug!", "Fixpoint.Debugging.notifyPreSolve", "Fixpoint.Debugging.notifyPostSolve", "Fixpoint.Debugging.notifyPreInterpret", "Assert.eq")
+
+  private def toReadableFunction(sym: Symbol.DefnSym): String = {
+    sym.toString.takeWhile(c => c.toString != Flix.Delimiter)
+  }
+
+  /**
+    * Performs occurrence analysis on the given AST `root`.
+    */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    val defs = visitDefs(root.defs) // visitDefs calls parMap internally and has additional handling
+    OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+  }
+
+  /**
+    * Performs occurrence analysis on every entry in `defs0` in parallel.
+    * Decorates each Def with occurrence information, i.e., how it appears in the program or if it is unused.
+    */
+  private def visitDefs(defs0: Map[DefnSym, OccurrenceAst.Def])(implicit flix: Flix): Map[DefnSym, OccurrenceAst.Def] = {
+    val (ds, os) = ParOps.parMap(defs0.values)(visitDef).unzip
+
+    // Combine all `defOccurrences` into one map.
+    val defOccur = combineAll(os)
+
+    // Updates the occurrence of every `def` in `ds` based on the occurrence found in `defOccur`.
+    ds.foldLeft(Map.empty[DefnSym, OccurrenceAst.Def]) {
+      case (macc, defn) =>
+        val occur = if (DangerousFunctions.contains(toReadableFunction(defn.sym))) DontInlineAndDontRewrite else defOccur.getOrElse(defn.sym, Dead)
+        val newContext = defn.context.copy(occur = occur)
+        val defWithContext = defn.copy(context = newContext)
+        macc + (defn.sym -> defWithContext)
+    }
+  }
+
+  /**
+    * Performs occurrence analysis on `defn`.
+    */
+  private def visitDef(defn0: OccurrenceAst.Def): (OccurrenceAst.Def, OccurInfo) = {
+
+    /**
+      * Returns true if `expr0` is a function call.
+      */
+    def isDirectCall(expr0: OccurrenceAst.Expr): Boolean = expr0 match {
+      case OccurrenceAst.Expr.ApplyDef(_, _, _, _, _, _) => true
+      case _ => false
+    }
+
+    /**
+      * Returns true if `def0` occurs in `occurInfo`.
+      */
+    def isSelfRecursive(occurInfo: OccurInfo): Boolean = occurInfo.defs.get(defn0.sym) match {
+      case None => false
+      case Some(o) => o match {
+        case Occur.Dead => false
+        case Occur.Once => true
+        case Occur.OnceInLambda => true
+        case Occur.OnceInLocalDef => true
+        case Occur.Many => true
+        case Occur.ManyBranch => true
+        case Occur.DontInline => false
```

Comment:
mark as `true` anyway?

Reply:
It is not totally clear to me how these Dont objects relate to occurrence. Conceptually it should be (Occur, Danger) since they are two different things. Do you think its practically okay to separate the two? Otherwise I think the docs for DontX need to say "DontX implies that it is not Dead" fx

Reply:
Hmm I see what you mean. The Haskell inliner also has a `DontInline` constructor which also tells the inliner hands off. So it's more of a "whatever value you return does not matter since `DontInline` takes precedence. So if a function marked with `DontInline` gets inlined, then we have a bug. Maybe we should just ignore this (set it to `true`) and check on every inlining decision that its occurrence information is not `DontInline`. If it is, then it should crash. However, they way it is written now also enforces that without crashing.